bentoml/_internal/server/grpc/interceptors/prometheus.py

```diff
@@ -0,0 +1,116 @@
+
+logger = logging.getLogger(__name__)
+START_TIME_VAR: contextvars.ContextVar[float] = contextvars.ContextVar("START_TIME_VAR")
+
+
+class PrometheusServerInterceptor(aio.ServerInterceptor):
+    """
+    An async interceptor for prometheus metrics
+    """
+
+    @inject
+    def __init__(
+        self,
+        bento_service: Service,
+        metrics_client: PrometheusClient = Provide[BentoMLContainer.metrics_client],
+    ):
+        self.metrics_client = metrics_client
+
+        self.service_name = bento_service.name
+        # a valid tag name may includes invalid characters, so we need to escape them
+        # ref: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+        self.service_name = self.service_name.replace("-", ":").replace(".", "::")
+
+        self.service_version = (
+            bento_service.tag.version if bento_service.tag is not None else ""
+        )
+
+        self.metrics_request_duration = metrics_client.Histogram(
+            name=f"{self.service_name}_request_duration_seconds",
+            documentation=f"{self.service_name} API GRPC request duration in seconds",
+            labelnames=["api_name", "service_version", "grpc_response_code"],
+        )
+        self.metrics_request_total = metrics_client.Counter(
+            name=f"{self.service_name}_request_total",
+            documentation="Total number of GRPC requests",
+            labelnames=["api_name", "service_version", "grpc_response_code"],
+        )
+        self.metrics_request_in_progress = metrics_client.Gauge(
+            name=f"{self.service_name}_request_in_progress",
+            documentation="Total number of GRPC requests in progress now",
+            labelnames=["api_name", "service_version"],
+        )
+
+    async def intercept_service(
+        self,
+        continuation: t.Callable[[HandlerCallDetails], t.Awaitable[RpcMethodHandler]],
+        handler_call_details: HandlerCallDetails,
+    ) -> RpcMethodHandler:
+        handler = await continuation(handler_call_details)
+
+        if handler and (handler.response_streaming or handler.request_streaming):
+            return handler
+
+        def wrapper(behaviour: AsyncHandlerMethod[Response]):
+            START_TIME_VAR.set(default_timer())
+
+            @functools.wraps(behaviour)
+            async def new_behavior(
+                request: Request, context: BentoServicerContext
+            ) -> Response | t.Awaitable[Response]:
+                api_name = request.api_name
+
+                # instrument request total count
+                self.metrics_request_total.labels(
+                    api_name=api_name,
+                    service_version=self.service_version,
+                    grpc_response_code=context.code(),
+                ).inc()
+
+                # instrument request duration
+                assert START_TIME_VAR.get() != 0
+                total_time = max(default_timer() - START_TIME_VAR.get(), 0)
```

Comment:
Timer should wrap the rpc calls, so I believe this is not correct.
Use `try..finally` loop.